feature/s3/transfermanager/putobject.go

```diff
@@ -0,0 +1,772 @@
+	// Copy ClientOptions
+	clientOptions := make([]func(*s3.Options), 0, len(c.options.PutClientOptions)+1)
+	clientOptions = append(clientOptions, func(o *s3.Options) {
+		o.APIOptions = append(o.APIOptions,
+			middleware.AddSDKAgentKey(middleware.FeatureMetadata, userAgentKey),
+			addFeatureUserAgent, // yes, there are two of these
+			func(s *smithymiddleware.Stack) error {
+				return s.Finalize.Insert(&setS3ExpressDefaultChecksum{}, "ResolveEndpointV2", smithymiddleware.After)
+			},
+		)
+	})
+	clientOptions = append(clientOptions, c.options.PutClientOptions...)
+	i.cfg.options.PutClientOptions = clientOptions
+	for _, opt := range opts {
+		opt(&i.cfg.options)
+	}
+
+	return i.upload()
+}
+
+type uploader struct {
+	ctx context.Context
+	cfg Client
+
+	in *PutObjectInput
+
+	//totalSize int64 // set to -1 if the size is not known
+}
+
+func (u *uploader) upload() (*PutObjectOutput, error) {
+	if err := u.init(); err != nil {
+		return nil, fmt.Errorf("unable to initialize upload: %w", err)
+	}
+	defer u.cfg.options.partPool.Close()
+
+	r, _, cleanUp, err := u.nextReader()
+
+	if err == io.EOF {
+		return u.singleUpload(r, cleanUp)
+	} else if err != nil {
+		cleanUp()
+		return nil, err
+	}
+
+	mu := multiUploader{
+		uploader: u,
+	}
+	return mu.upload(r, cleanUp)
+}
+
+func (u *uploader) init() error {
+	if err := validateSupportedARNType(aws.ToString(u.in.Bucket)); err != nil {
+		return err
+	}
+
+	o := &u.cfg.options
+	if o.S3 == nil {
+		cfg, err := config.LoadDefaultConfig(context.TODO())
+		if err != nil {
+			return fmt.Errorf("error while creating default s3 cfg: %q", err)
+		}
+		o.S3 = s3.NewFromConfig(cfg)
+	}
+
+	if o.Concurrency == 0 {
+		o.Concurrency = DefaultTransferConcurrency
+	}
+	if o.PartSizeBytes == 0 {
+		o.PartSizeBytes = DefaultPartSizeBytes
+	} else if o.PartSizeBytes < DefaultPartSizeBytes {
+		return fmt.Errorf("part size must be at least %d bytes", DefaultPartSizeBytes)
+	}
+	if o.MaxUploadParts == 0 {
+		o.MaxUploadParts = DefaultMaxUploadParts
+	} else if o.MaxUploadParts > DefaultMaxUploadParts {
+		return fmt.Errorf("max upload parts must be at most %d bytes", DefaultMaxUploadParts)
+	}
+	if err := u.initSize(); err != nil {
+		return err
+	}
+	if o.ChecksumAlgorithm == "" {
+		o.ChecksumAlgorithm = types.ChecksumAlgorithmCrc32
+	}
+
+	// If PartSize was changed or partPool was never setup then we need to allocated a new pool
+	// so that we return []byte slices of the correct size
+	poolCap := o.Concurrency + 1
+	if o.partPool == nil || o.partPool.SliceSize() != o.PartSizeBytes {
+		o.partPool = newByteSlicePool(o.PartSizeBytes)
+		o.partPool.ModifyCapacity(poolCap)
+	} else {
+		o.partPool = &returnCapacityPoolCloser{byteSlicePool: o.partPool}
+		o.partPool.ModifyCapacity(poolCap)
+	}
+
+	return nil
+}
+
+// initSize tries to detect the total stream size, setting u.totalSize. If
+// the size is not known, totalSize is set to -1.
+func (u *uploader) initSize() error {
```

Comment:
I don't really see how this is necessary. We should just keep pulling bytes off of the input (into slices from the pool) until we run out.